tests/csapi/apidoc_sendtodevice_test.go

```diff
@@ -0,0 +1,196 @@
+
+		// Advance the sync token, we should get one event
+		t.Logf("Syncing with %s", nextBatch)
+		prevNextBatch := nextBatch
+		nextBatch = bob.MustSyncUntil(t, client.SyncReq{Since: nextBatch}, verifyToDeviceResp(t, bob, checks, 1))
+
+		// act as if the sync "failed", so we're using the previous next batch
+		t.Logf("Syncing with %s", prevNextBatch)
+		bob.MustSyncUntil(t, client.SyncReq{Since: prevNextBatch}, verifyToDeviceResp(t, bob, checks, 1))
+
+		// advance since, to clear previously received events
+		t.Logf("Syncing with %s", nextBatch)
+		bob.MustSyncUntil(t, client.SyncReq{Since: nextBatch}, verifyToDeviceResp(t, bob, checks, 0))
+
+		// since we advanced "since", this should return no events anymore
+		t.Logf("Syncing with %s", prevNextBatch)
+		bob.MustSyncUntil(t, client.SyncReq{Since: prevNextBatch}, verifyToDeviceResp(t, bob, checks, 0))
+	})
+
+	t.Run("sendToDevice messages are ordered by arrival", func(t *testing.T) {
+		for i := 0; i < 10; i++ {
+			// Create a m.room_key_request sendToDevice message
+			reqBody := client.WithJSONBody(t, map[string]interface{}{
+				"messages": map[string]interface{}{
+					bob.UserID: map[string]interface{}{
+						bob.DeviceID: map[string]interface{}{
+							"action":               "request",
+							"requesting_device_id": alice.DeviceID,
+							"request_id":           i,
+						},
+					},
+				},
+			})
+			time.Sleep(time.Millisecond * 100) // Wait a bit for the server to process all messages
+
+			txnID := strconv.Itoa(10 + i)
+			alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "sendToDevice", "m.room_key_request", txnID}, reqBody)
+		}
+
+		verifyOrdering := func() match.JSON {
+			var i int64
+			return func(body []byte) error {
+				if gotID := gjson.GetBytes(body, "content.request_id").Int(); gotID != i {
+					return fmt.Errorf("expected request_id to be %d, got %d", i, gotID)
+				}
+				i++
+				return nil
+			}
+		}()
+
+		bob.MustSyncUntil(t, client.SyncReq{}, verifyToDeviceResp(t, bob, []match.JSON{verifyOrdering}, 10))
+	})
+
+}
+
+func verifyToDeviceResp(t *testing.T, user *client.CSAPI, checks []match.JSON, wantMessageCount int64) client.SyncCheckOpt {
+	return func(userID string, syncResp gjson.Result) error {
+		t.Helper()
+		toDevice := syncResp.Get("to_device.events")
+		if !toDevice.Exists() {
+			if wantMessageCount > 0 && len(toDevice.Array()) == 0 {
+				return fmt.Errorf("no to_device.events found: %s", syncResp.Get("to_device").Raw)
+			}
+		}
+		if count := len(toDevice.Array()); int64(count) != wantMessageCount {
+			t.Fatalf("(%s - %s) expected %d to_device.events, got %d - %v", user.UserID, user.DeviceID, wantMessageCount, count, toDevice.Raw)
+		}
+		for _, message := range toDevice.Array() {
+			for _, check := range checks {
+				if err := check([]byte(message.Raw)); err != nil {
+					return fmt.Errorf("%v: %s", err, message.Raw)
```

Comment:
Something like `SendToDeviceSynced` could work but would be tricky because this would mean using `/sync` to verify it has arrived (which would be on the receiver as well, which is different to say `SendEventSynced`) - if you advance beyond that position then the event will be lost. It's a bit too error prone I think.

This current implementation is a bit too brittle because it will immediately fail the test if the /sync response does not have `wantMessageCount`, which is probably why the `sleep()` was added to ensure all 10 to-device messages would be waiting for the user **in a single sync response**. This is brittle because you cannot know when these events will be ready, and you simply need to keep trying until you reach a sensible timeout value. This is exactly what all the sync methods do, and we should be making use of them more - i.e take `SyncToDeviceHas` from https://github.com/matrix-org/complement/pull/439 and then do:

```go
var i int64
bob.MustSyncUntil(t, client.SyncReq{}, client.SyncToDeviceHas(
    func(msg gjson.Result) bool {
        gotReqID := msg.Get("content.request_id").Int()
        if gotReqID == i {
            i++ // we have the next request ID, look for another
        } else {
            // if we see any other request ID, whine about it e.g out-of-order requests
            t.Errorf("unexpected to-device request_id: %d", gotReqID)
        }
        return i == 10 // terminate when we have seen all requests in order
    }
))
```

This will work for both cases where:
 - /sync returns all 10 events.
 - /sync returns 7 events, then the next one returns 2, then the next one returns 1.

This will also fail gracefully if:
 - events are missing. We time out after 5s, no sleeps needed.
 - events arrive out-of-order. We get a nice error message.